sklearn/decomposition/tests/test_fastica.py

```diff
@@ -265,10 +298,20 @@ def test_fit_transform():
         (False, 10, (10, 10)),
     ],
 )
-def test_inverse_transform(whiten, n_components, expected_mixing_shape):
+def test_inverse_transform(
+    whiten, n_components, expected_mixing_shape, global_random_seed
+):
+    # XXX: casting X with `.astype(global_dtype)` reveals a rare numerical
+    # stability problem of FastICA with float32 bit data. I can triggers
+    # NaN values somewhere in the computation, only for some random seeds.
+
+    # XXX: furthermore, even when we do not hit the NaN issue, the
+    # assert_allclose can only pass with rtol values significantly larger than
+    # the usual 1e-4 we expect for float32 data.
```

Comment:
Similar comment here. The nans can appear at fit time but even when it's not the case, rtol would need to be large (larger than 1e-1) for some seeds.

I think we can safely say that our implementation is not numerically stable enough to work with float32 data without upcasting at this point.